Windows - Support for -Clink-dead-code

I have to enable the -Clink-dead-code in order to generate code coverage for one of my projects. However, there seems to be a problem coming from skia-safe when linking with that flag. I'm following the instructions provided by [grcov](https://github.com/mozilla/grcov). I'm able to build correctly if I remove that flag.

**Build command**
```
export CARGO_INCREMENTAL=0
export RUSTFLAGS="-Zprofile -Ccodegen-units=1 -Copt-level=0 -Clink-dead-code -Coverflow-checks=off -Zno-landing-pads"
cargo build
cargo test
```

**Example of linking errors**
```
"public: void __cdecl SkString::`vbase destructor'(void)" (??_DSkString@@QEAAXXZ) referenced in function _ZN13skia_bindings8bindings8SkString8destruct17h2e
8082931f49dff6E libskia_bindings-9d5bde54cbd3a647.rlib(skia_bindings-9d5bde54cbd3a647.skia_bindings.4thq6d2u-cgu.0.rcgu.o) : error LNK2019: unresolved external symbol "public: void __cdecl SkDeque::`vbase destructor'(void)" (??_DSkDeque@@QEAAXXZ) referenced in function _ZN13skia_bindings8bindings7SkDeque8destruct17haf8c3
c3ff4355596E libskia_bindings-9d5bde54cbd3a647.rlib(skia_bindings-9d5bde54cbd3a647.skia_bindings.4thq6d2u-cgu.0.rcgu.o) : error LNK2019: unresolved external symbol "public: void __cdecl SkPaint::`vbase destructor'(void)" (??_DSkPaint@@QEAAXXZ) referenced in function _ZN13skia_bindings8bindings7SkPaint8destruct17h43ebe
74856ac54bfE
```

**Generated linker command:**
```
error: linking with `link.exe` failed: exit code: 1120 |
  = note: "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\VC\\Tools\\MSVC\\14.23.28105\\bin\\HostX64\\x64\\link.exe" "/NOLOGO" "/NXCOMPAT" "/LIBPATH:C:\\Users\\rachidox\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib" "C:\\Users\\rachidox\\Home\\g
ithub\\universe-skia-poc\\skia-safe-poc\\target\\debug\\deps\\skia_safe_poc-a06c1188f6c2379f.skia_safe_poc.4wof72ks-cgu.0.rcgu.o" "/OUT:C:\\Users\\rachidox\\H
ome\\github\\universe-skia-poc\\skia-safe-poc\\target\\debug\\deps\\skia_safe_poc-a06c1188f6c2379f.exe" "/DEBUG" "/NATVIS:C:\\Users\\rachidox\\.rustup\\toolch
ains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\intrinsic.natvis" "/NATVIS:C:\\Users\\rachidox\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\
lib\\rustlib\\etc\\liballoc.natvis" "/NATVIS:C:\\Users\\rachidox\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\libcore.natvis" "/NA
TVIS:C:\\Users\\rachidox\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\libstd.natvis" "/LIBPATH:C:\\Users\\rachidox\\Home\\github\\
universe-skia-poc\\skia-safe-poc\\target\\debug\\deps" "/LIBPATH:C:\\Users\\rachidox\\Home\\github\\universe-skia-poc\\skia-safe-poc\\target\\debug\\build\\skia-bindings-31807756e9d02dd6\\out\\skia" "/LIBPATH:C:\\Users\\rachidox\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-m
svc\\lib" "/LIBPATH:C:\\Users\\rachidox\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib" "test-16d7664c2b987d4
5.dll.lib" "C:\\Users\\rachidox\\Home\\github\\universe-skia-poc\\skia-safe-poc\\target\\debug\\deps\\libskia_safe-7af3055b17345ac2.rlib" "C:\\Users\\rachidox
\\Home\\github\\universe-skia-poc\\skia-safe-poc\\target\\debug\\deps\\libskia_bindings-9d5bde54cbd3a647.rlib" "C:\\Users\\rachidox\\Home\\github\\universe-sk
ia-poc\\skia-safe-poc\\target\\debug\\deps\\liblazy_static-03a16b23802d8762.rlib" "C:\\Users\\rachidox\\Home\\github\\universe-skia-poc\\skia-safe-poc\\target
\\debug\\deps\\libbitflags-e9b503ab5e675abc.rlib" "C:\\Users\\rachidox\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libprofiler_builtins-d6f4b4bfe2d2890a.rlib" "/LIBPATH:C:\\Users\\rachidox\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64
-pc-windows-msvc\\lib" "std-597c5a05d736e633.dll.lib" "C:\\Users\\rachidox\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcompiler_builtins-94f7006c25e00caf.rlib" "kernel32.lib" "usp10.lib" "ole32.lib" "user32.lib" "gdi32.lib" "fontsub.lib" "advapi32.lib" "ws2_32
.lib" "userenv.lib" "msvcrt.lib"
  = note: libskia_bindings-9d5bde54cbd3a647.rlib(skia_bindings-9d5bde54cbd3a647.skia_bindings.4thq6d2u-cgu.0.rcgu.o) : error LNK2019: unresolved external symb
ol "public: void __cdecl SkString::`vbase destructor'(void)" (??_DSkString@@QEAAXXZ) referenced in function _ZN13skia_bindings8bindings8SkString8destruct17h2e
8082931f49dff6E
          libskia_bindings-9d5bde54cbd3a647.rlib(skia_bindings-9d5bde54cbd3a647.skia_bindings.4thq6d2u-cgu.0.rcgu.o) : error LNK2019: unresolved external symb
ol "public: void __cdecl SkDeque::`vbase destructor'(void)" (??_DSkDeque@@QEAAXXZ) referenced in function _ZN13skia_bindings8bindings7SkDeque8destruct17haf8c3
c3ff4355596E
```

Thanks for this great Crate. Actually I tried to link against the fix branch, there are less issues however I still see these:
```
note: libskia_bindings-6f59f90af691f8cb.rlib(skia_bindings-6f59f90af691f8cb.skia_bindings.9uzl4pl8-cgu.0.rcgu.o) : error LNK2019: unresolved exter
nal symbol "public: class SkDeferredDisplayListPriv __cdecl SkDeferredDisplayList::priv(void)" (?priv@SkDeferredDisplayList@@QEAA?AVSkDeferredDisplayL
istPriv@@XZ) referenced in function _ZN13skia_bindings8bindings21SkDeferredDisplayList5priv_17heedb8d99ab718c63E
          libskia_bindings-6f59f90af691f8cb.rlib(skia_bindings-6f59f90af691f8cb.skia_bindings.9uzl4pl8-cgu.0.rcgu.o) : error LNK2019: unresolved exter
nal symbol "public: class SkDeferredDisplayListPriv const __cdecl SkDeferredDisplayList::priv(void)const " (?priv@SkDeferredDisplayList@@QEBA?BVSkDefe
rredDisplayListPriv@@XZ) referenced in function _ZN13skia_bindings8bindings21SkDeferredDisplayList5priv117h29bba020676b96d9E
          libskia_bindings-6f59f90af691f8cb.rlib(skia_bindings-6f59f90af691f8cb.skia_bindings.9uzl4pl8-cgu.0.rcgu.o) : error LNK2019: unresolved exter
nal symbol "public: static class sk_sp<class SkImage> __cdecl SkImage::MakeTextureFromCompressed(class GrContext *,class sk_sp<class SkData>,int,int,e
num SkImage::CompressionType,enum GrMipMapped,enum GrProtected)" (?MakeTextureFromCompressed@SkImage@@SA?AV?$sk_sp@VSkImage@@@@PEAVGrContext@@V?$sk_sp
@VSkData@@@@HHW4CompressionType@1@W4GrMipMapped@@W4GrProtected@@@Z) referenced in function _ZN13skia_bindings8bindings7SkImage25MakeTextureFromCompres
sed17h211737b4cf3addd3E
          libskia_bindings-6f59f90af691f8cb.rlib(skia_bindings-6f59f90af691f8cb.skia_bindings.9uzl4pl8-cgu.0.rcgu.o) : error LNK2019: unresolved exter
nal symbol "public: static class sk_sp<class SkImage> __cdecl SkImage::MakeCrossContextFromPixmap(class GrContext *,class SkPixmap const &,bool,bool)"
 (?MakeCrossContextFromPixmap@SkImage@@SA?AV?$sk_sp@VSkImage@@@@PEAVGrContext@@AEBVSkPixmap@@_N2@Z) referenced in function _ZN13skia_bindings8bindings
7SkImage26MakeCrossContextFromPixmap17hbbb63a1fd4168b0bE
          libskia_bindings-6f59f90af691f8cb.rlib(skia_bindings-6f59f90af691f8cb.skia_bindings.9uzl4pl8-cgu.0.rcgu.o) : error LNK2019: unresolved exter
nal symbol "public: static class sk_sp<class SkImage> __cdecl SkImage::MakeFromYUVATextures(class GrContext *,enum SkYUVColorSpace,class GrBackendText
ure const * const,struct SkYUVAIndex const * const,struct SkISize,enum GrSurfaceOrigin,class sk_sp<class SkColorSpace>)" (?MakeFromYUVATextures@SkImag
e@@SA?AV?$sk_sp@VSkImage@@@@PEAVGrContext@@W4SkYUVColorSpace@@QEBVGrBackendTexture@@QEBUSkYUVAIndex@@USkISize@@W4GrSurfaceOrigin@@V?$sk_sp@VSkColorSpa
ce@@@@@Z) referenced in function _ZN13skia_bindings8bindings7SkImage20MakeFromYUVATextures17hf8af6dad6955c7e0E
          libskia_bindings-6f59f90af691f8cb.rlib(skia_bindings-6f59f90af691f8cb.skia_bindings.9uzl4pl8-cgu.0.rcgu.o) : error LNK2019: unresolved exter
nal symbol "public: static class sk_sp<class SkImage> __cdecl SkImage::MakeFromYUVAPixmaps(class GrContext *,enum SkYUVColorSpace,class SkPixmap const
 * const,struct SkYUVAIndex const * const,struct SkISize,enum GrSurfaceOrigin,bool,bool,class sk_sp<class SkColorSpace>)" (?MakeFromYUVAPixmaps@SkImag
e@@SA?AV?$sk_sp@VSkImage@@@@PEAVGrContext@@W4SkYUVColorSpace@@QEBVSkPixmap@@QEBUSkYUVAIndex@@USkISize@@W4GrSurfaceOrigin@@_N6V?$sk_sp@VSkColorSpace@@@
@@Z) referenced in function _ZN13skia_bindings8bindings7SkImage19MakeFromYUVAPixmaps17hb278117b416493a1E
          libskia_bindings-6f59f90af691f8cb.rlib(skia_bindings-6f59f90af691f8cb.skia_bindings.9uzl4pl8-cgu.0.rcgu.o) : error LNK2019: unresolved exter
nal symbol "public: static class sk_sp<class SkSurface> __cdecl SkSurface::MakeFromBackendTexture(class GrContext *,class SkSurfaceCharacterization co
nst &,class GrBackendTexture const &,void (__cdecl*)(void *),void *)" (?MakeFromBackendTexture@SkSurface@@SA?AV?$sk_sp@VSkSurface@@@@PEAVGrContext@@AE
BVSkSurfaceCharacterization@@AEBVGrBackendTexture@@P6AXPEAX@Z3@Z) referenced in function _ZN13skia_bindings8bindings9SkSurface23MakeFromBackendTexture
117hf9f408d833e946eaE
          C:\Users\rachidox\Home\github\universe-skia-poc\skia-safe-poc\target\debug\deps\skia_safe_poc-a62f84347b29b927.exe : fatal error LNK1120: 7
unresolved externals
```

Here an update about the linker errors I was able to reproduce:

Compiling with:
```bash
export CARGO_INCREMENTAL=0
export RUSTFLAGS="-Ccodegen-units=1 -Copt-level=0 -Clink-dead-code -Coverflow-checks=off"
cargo build -vv
```


the Windows build with the Branch `link-dead-code` (with all features enabled, "gl", "vulkan", and "textlayout") works so without any linker errors. But I had no luck to generate the bindings with the nightly compiler, somehow the binding generation just stops.

The macOS (LLVM 10) build with all features (the ones above + "metal") fails to link with one linker error:

```
  = note: Undefined symbols for architecture x86_64:
            "_CFMutableAttributedStringGetTypeID", referenced from:
                _$LT$core_foundation..attributed_string..CFMutableAttributedString$u20$as$u20$core_foundation..base..TCFType$GT$::type_id::h2bbb20d9af45e39f in libcore_foundation-4b8d866b003bda1a.rlib(core_foundation-4b8d866b003bda1a.core_foundation.7u55rrq4-cgu.0.rcgu.o)
                _$LT$core_foundation..attributed_string..CFMutableAttributedString$u20$as$u20$core_foundation..base..TCFType$GT$::type_id::h7a7d878dcafc09fc in libcore_foundation-62a6a57942bad80d.rlib(core_foundation-62a6a57942bad80d.core_foundation.en1q3mex-cgu.0.rcgu.o)
          ld: symbol(s) not found for architecture x86_64
```
which is probably related to [this issue](https://github.com/servo/core-foundation-rs/pull/357) but may not happen with other executables built with skia-safe.

I am considering to merge the branch as "better than before" and close this issue because there is currently no action I am able to proceed with. Please reopen or create a new issue if new information comes up.

--- a/skia-bindings/build_support/skia.rs
+++ b/skia-bindings/build_support/skia.rs
@@ -1,6 +1,7 @@
 //! Full build support for the Skia library, SkiaBindings library and bindings.rs file.
 
 use crate::build_support::{android, binaries, cargo, clang, ios, llvm, vs, xcode};
+use bindgen::{CodegenConfig, EnumVariation};
 use cc::Build;
 use std::path::{Path, PathBuf};
 use std::process::{Command, Stdio};
@@ -543,6 +544,34 @@ fn bindgen_gen(build: &FinalBuildConfiguration, current_dir: &Path, output_direc
         .constified_enum("SkCanvas_SaveLayerFlagsSet")
         .constified_enum("GrVkAlloc_Flag")
         .constified_enum("GrGLBackendState")
+        // not used:
+        .blacklist_type("SkPathRef_Editor")
+        .blacklist_function("SkPathRef_Editor_Editor")
+        // private types that pull in inline functions that cannot be linked:
+        // https://github.com/rust-skia/rust-skia/issues/318
+        .raw_line("pub enum GrContext_Base {}")
+        .blacklist_type("GrContext_Base")
+        .blacklist_function("GrContext_Base_.*")
+        .raw_line("pub enum GrImageContext {}")
+        .blacklist_type("GrImageContext")
+        .raw_line("pub enum GrImageContextPriv {}")
+        .blacklist_type("GrImageContextPriv")
+        .raw_line("pub enum GrContextThreadSafeProxy {}")
+        .blacklist_type("GrContextThreadSafeProxy")
+        .raw_line("pub enum GrContextThreadSafeProxyPriv {}")
+        .blacklist_type("GrContextThreadSafeProxyPriv")
+        .raw_line("pub enum GrRecordingContext {}")
+        .blacklist_type("GrRecordingContext")
+        .raw_line("pub enum GrRecordingContextPriv {}")
+        .blacklist_type("GrRecordingContextPriv")
+        .raw_line("pub enum GrContextPriv {}")
+        .blacklist_type("GrContextPriv")
+        .raw_line("fn GrContext_priv(_: &GrContext) -> GrContextPriv { panic!(\"unexpected\") }")
+        .raw_line("fn GrContext_priv1(_: &GrContext) -> GrContextPriv { panic!(\"unexpected\") }")
+        .blacklist_function("GrContext_priv.*")
+        .raw_line("fn SkDeferredDisplayList_priv(_: &SkDeferredDisplayList) -> SkDeferredDisplayListPriv { panic!(\"unexpected\") }")
+        .raw_line("fn SkDeferredDisplayList_priv1(_: &SkDeferredDisplayList) -> SkDeferredDisplayListPriv { panic!(\"unexpected\") }")
+        .blacklist_function("SkDeferredDisplayList_priv.*")
         // Vulkan reexports that got swallowed by making them opaque.
         // (these can not be whitelisted by a extern "C" function)
         .whitelist_type("VkPhysicalDeviceFeatures")
@@ -550,6 +579,13 @@ fn bindgen_gen(build: &FinalBuildConfiguration, current_dir: &Path, output_direc
         // misc
         .whitelist_var("SK_Color.*")
         .whitelist_var("kAll_GrBackendState")
+        //
+        // don't generate destructors: https://github.com/rust-skia/rust-skia/issues/318
+        .with_codegen_config({
+            let mut config = CodegenConfig::default();
+            config.remove(CodegenConfig::DESTRUCTORS);
+            config
+        })
         //
         .use_core()
         .clang_arg("-std=c++17")
@@ -689,7 +725,6 @@ const OPAQUE_TYPES: &[&str] = &[
     "SkDeferredDisplayList_PendingPathsMap",
     // Types for which a bindgen layout is wrong causing types that contain
     // fields of them to fail their layout test.
-
     // Windows:
     "std::atomic",
     "std::function",
@@ -699,11 +734,7 @@ const OPAQUE_TYPES: &[&str] = &[
     // Ubuntu 18 LLVM 6: all types derived from SkWeakRefCnt
     "SkWeakRefCnt",
     "GrContext",
-    "GrContextThreadSafeProxy",
-    "GrContext_Base",
     "GrGLInterface",
-    "GrImageContext",
-    "GrRecordingContext",
     "GrSurfaceProxy",
     "Sk2DPathEffect",
     "SkCornerPathEffect",
@@ -788,6 +819,11 @@ const OPAQUE_TYPES: &[&str] = &[
     "GrShaderCaps",
     // m81: yet experimental
     "SkM44",
+    // more stuff we don't need that was tracked down fixing:
+    // https://github.com/rust-skia/rust-skia/issues/318
+    // referred from SkPath, but not used:
+    "SkPathRef",
+    "SkMutex",
 ];
 
 const BLACKLISTED_TYPES: &[&str] = &[
@@ -1179,7 +1215,6 @@ mod prerequisites {
     }
 }
 
-use bindgen::EnumVariation;
 pub use definitions::{Definition, Definitions};
 
 pub(crate) mod definitions {

--- a/skia-bindings/build_support/skia.rs
+++ b/skia-bindings/build_support/skia.rs
@@ -35,7 +35,6 @@ impl Default for BuildConfiguration {
         BuildConfiguration {
             on_windows: cargo::host().is_windows(),
             skia_debug,
-            keep_inline_functions: true,
             features: Features {
                 gl: cfg!(feature = "gl"),
                 vulkan: cfg!(feature = "vulkan"),
@@ -58,10 +57,6 @@ pub struct BuildConfiguration {
 
     /// Build Skia in a debug configuration?
     skia_debug: bool,
-
-    /// Configure Skia builds to keep inline functions to
-    /// prevent linker errors.
-    keep_inline_functions: bool,
 
     /// The Skia feature set to compile.
     features: Features,

